File: CarHomeWidget/src/com/anod/car/home/incar/ModeService.java
package com.anod.car.home.incar;

import android.app.Notification;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.IBinder;
import android.telephony.PhoneStateListener;
import android.telephony.TelephonyManager;
import android.util.Log;
import android.view.View;
import android.widget.RemoteViews;

import com.anod.car.home.Provider;
import com.anod.car.home.R;
import com.anod.car.home.ShortcutPendingIntent;
import com.anod.car.home.model.NotificationShortcutsModel;
import com.anod.car.home.model.ShortcutInfo;
import com.anod.car.home.prefs.PreferencesStorage;
import com.anod.car.home.prefs.preferences.InCar;
import com.anod.car.home.utils.Utils;

public class ModeService extends Service{
	private static final String PREFIX_NOTIF = "notif";
	private ModePhoneStateListener mPhoneListener;
	private int[] sBtnIds = {
		R.id.btn0,
		R.id.btn1,
		R.id.btn2,
		R.id.btn3
	};
	private boolean mForceState = false;
	public static boolean sInCarMode = false;
	private static final int NOTIFICATION_ID = 1;
	public static String EXTRA_MODE = "extra_mode";
	public static String EXTRA_FORCE_STATE = "extra_force_state";
	public static final int MODE_SWITCH_OFF = 1;
	public static final int MODE_SWITCH_ON = 0;

	@Override
	public void onCreate() {
		
		super.onCreate();
	}

	private Notification createNotification() {
		Intent notificationIntent = new Intent(this, ModeService.class);
		notificationIntent.putExtra(EXTRA_MODE, MODE_SWITCH_OFF);
    	Uri data = Uri.parse("com.anod.car.home.pro://mode/0/");
    	notificationIntent.setData(data);
		
		PendingIntent contentIntent = PendingIntent.getService(this, 0, notificationIntent, 0);

		Notification notification = new Notification();
		notification.flags |= Notification.FLAG_ONGOING_EVENT;
		notification.icon = R.drawable.ic_stat_incar;
		if (Utils.IS_HONEYCOMB_OR_GREATER) {
			RemoteViews contentView = createShortcuts();
			notification.contentIntent = contentIntent;
			notification.contentView = contentView;
		} else {
			String notifTitle=getResources().getString(R.string.incar_mode_enabled);
			String notifText=getResources().getString(R.string.click_to_disable);
			notification.setLatestEventInfo(this, notifTitle, notifText, contentIntent);
		}
		return notification;
	}

	private RemoteViews createShortcuts() {
		RemoteViews contentView = new RemoteViews(getPackageName(), R.layout.notification);
		NotificationShortcutsModel model = new NotificationShortcutsModel(this);
		model.init();
		boolean viewGone = true;
		for (int i = 0; i < model.getCount(); i++) {
			ShortcutInfo info = model.getShortcut(i);
			int resId = sBtnIds[i];
			if (info == null) {
				contentView.setViewVisibility(resId, (viewGone) ? View.GONE : View.INVISIBLE);
			} else {
				viewGone = false;
				contentView.setImageViewBitmap(resId, info.getIcon());
				PendingIntent pendingIntent = ShortcutPendingIntent.getShortcutPendingInent(info.intent, PREFIX_NOTIF, this, i);
				contentView.setOnClickPendingIntent(resId, pendingIntent);
			}
		}
		return contentView;
	}

	@Override
	public void onDestroy() {
		stopForeground(true);
		InCar prefs = PreferencesStorage.loadInCar(this);
		if (mForceState) {
			Handler.forceState(prefs, false);
		}
		Handler.switchOff(prefs,this);
        if (mPhoneListener != null) {
        	detachPhoneListener();
        }
		sInCarMode = false;
		requestWidgetsUpdate();
		super.onDestroy();
	}

	private void requestWidgetsUpdate() {
        Provider appWidgetProvider = Provider.getInstance();
        appWidgetProvider.performUpdate(this, null);
	}

	@Override
	public int onStartCommand(Intent intent, int flags, int startId)
	{
		// Tracking a bug
		if (intent == null) {
			Log.d("CarHomeWidget", "Intent is null...");
		}
		mForceState = intent.getBooleanExtra(EXTRA_FORCE_STATE, false);
		if (intent.getIntExtra(EXTRA_MODE, MODE_SWITCH_ON) == MODE_SWITCH_OFF) {
			stopSelf();
			return START_STICKY;
		}
		InCar prefs = PreferencesStorage.loadInCar(this);
		sInCarMode = true;
		if (mForceState) {
			Handler.forceState(prefs, true);
		}
		Handler.switchOn(prefs,this);
		handlePhoneListener(prefs);
		requestWidgetsUpdate();
		
		Notification notification = createNotification();
		startForeground(NOTIFICATION_ID, notification);

		// We want this service to continue running until it is explicitly
		// stopped, so return sticky.
		return START_STICKY;
	}

	private void handlePhoneListener(InCar prefs) {
    	if ( prefs.isAutoSpeaker() || 
    		!prefs.getAutoAnswer().equals(PreferencesStorage.AUTOANSWER_DISABLED)
    	) {
    		if (mPhoneListener == null) {
    			attachPhoneListener();
    		}
    		mPhoneListener.setActions(prefs.isAutoSpeaker(), prefs.getAutoAnswer());
    	} else {
            if (mPhoneListener != null) {
            	detachPhoneListener();
            }
    	}		
	}
	
	private void attachPhoneListener() {
		Log.d("CarHomeWidget", "Set phone listener");
		TelephonyManager tm = (TelephonyManager)this.getSystemService(Context.TELEPHONY_SERVICE);
		mPhoneListener = new ModePhoneStateListener(this);
		tm.listen(mPhoneListener, PhoneStateListener.LISTEN_CALL_STATE);
	}
	
	private void detachPhoneListener() {
    	Log.d("CarHomeWidget", "Remove phone listener");
    	TelephonyManager tm = (TelephonyManager)this.getSystemService(Context.TELEPHONY_SERVICE);
    	tm.listen(mPhoneListener, PhoneStateListener.LISTEN_NONE);
    	mPhoneListener.cancelActions();
    	mPhoneListener = null;
	}
	
	@Override
	public IBinder onBind(Intent arg0) {
		// TODO Auto-generated method stub
		return null;
	}
	
	
}

File: CarHomeWidget/src/main/java/com/anod/car/home/incar/ModeService.kt
package com.anod.car.home.incar

import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.IBinder
import android.os.PowerManager
import android.telephony.PhoneStateListener
import com.anod.car.home.appwidget.Provider
import com.anod.car.home.app.App
import com.anod.car.home.notifications.InCarModeNotification
import com.anod.car.home.notifications.TrialExpiredNotification
import com.anod.car.home.prefs.model.InCarInterface
import com.anod.car.home.prefs.model.InCarStorage
import info.anodsplace.framework.AppLog
import com.anod.car.home.utils.Version

class ModeService : Service() {
    private var phoneListener: ModePhoneStateListener? = null
    private val modeHandler: ModeHandler by lazy { App.provide(this).handler }
    private var forceState: Boolean = false

    override fun onDestroy() {
        stopForeground(true)

        val prefs = InCarStorage.load(this)
        if (forceState) {
            ModeDetector.forceState(prefs, false)
        }
        ModeDetector.switchOff(prefs, modeHandler)
        if (phoneListener != null) {
            detachPhoneListener()
        }

        sInCarMode = false
        requestWidgetsUpdate()

        super.onDestroy()
    }

    private fun requestWidgetsUpdate() {
        Provider.requestUpdate(this, intArrayOf())
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {

        val redelivered = flags and Service.START_FLAG_REDELIVERY == Service.START_FLAG_REDELIVERY
        AppLog.d("ModeService onStartCommand, sInCarMode = " + sInCarMode + ", redelivered = "
                + redelivered)

        if (intent == null) {
            AppLog.e("ModeService started without intent")
            stopSelf()
            return Service.START_NOT_STICKY
        }

        val mode = intent.getIntExtra(EXTRA_MODE, -1)
        if (mode == -1) {
            AppLog.e("ModeService, start mode is not correct")
            stopSelf()
            return Service.START_NOT_STICKY
        }
        if (mode == MODE_SWITCH_OFF) {
            stopSelf()
            return Service.START_NOT_STICKY
        }
        // mode == MODE_SWITCH_ON
        forceState = intent.getBooleanExtra(EXTRA_FORCE_STATE, false)

        val version = Version(this)
        if (version.isFreeAndTrialExpired) {
            TrialExpiredNotification.show(this)
            stopSelf()
            return Service.START_NOT_STICKY
        }

        val prefs = InCarStorage.load(this)
        sInCarMode = true
        if (forceState) {
            ModeDetector.forceState(prefs, true)
        }

        ModeDetector.switchOn(prefs, modeHandler)
        initPhoneListener(prefs)
        requestWidgetsUpdate()

        if (version.isFree) {
            version.increaseTrialCounter()
        }

        startForeground(InCarModeNotification.id, InCarModeNotification.create(version, this))

        // We want this service to continue running until it is explicitly
        // stopped, so return sticky.
        return Service.START_REDELIVER_INTENT
    }

    private fun initPhoneListener(prefs: InCarInterface) {
        if (prefs.isAutoSpeaker || prefs.autoAnswer != InCarInterface.AUTOANSWER_DISABLED) {
            if (phoneListener == null) {
                attachPhoneListener()
            }
            phoneListener!!.setActions(prefs.isAutoSpeaker, prefs.autoAnswer)
        } else {
            if (phoneListener != null) {
                detachPhoneListener()
            }
        }
    }

    private fun attachPhoneListener() {
        AppLog.d("Attach phone listener")
        val provider = App.provide(this)
        phoneListener = provider.modePhoneStateListener
        provider.telephonyManager.listen(phoneListener, PhoneStateListener.LISTEN_CALL_STATE)
    }

    private fun detachPhoneListener() {
        AppLog.d("Detach phone listener")
        val tm = App.provide(this).telephonyManager
        tm.listen(phoneListener, PhoneStateListener.LISTEN_NONE)
        phoneListener!!.cancelActions()
        phoneListener = null
    }

    override fun onBind(arg0: Intent): IBinder? {
        return null
    }

    companion object {

        private const val wakelockTag = "carhomewidget:wakelock"
        const val EXTRA_MODE = "extra_mode"
        const val EXTRA_FORCE_STATE = "extra_force_state"

        const val MODE_SWITCH_OFF = 1
        const val MODE_SWITCH_ON = 0

        var sInCarMode: Boolean = false

        @Volatile
        private var sLockStatic: PowerManager.WakeLock? = null

        @Synchronized
        private fun getLock(context: Context): PowerManager.WakeLock? {
            if (sLockStatic == null) {
                val mgr = context.getSystemService(Context.POWER_SERVICE) as PowerManager
                sLockStatic = mgr.newWakeLock(PowerManager.SCREEN_BRIGHT_WAKE_LOCK or PowerManager.ACQUIRE_CAUSES_WAKEUP, wakelockTag)
            }

            return sLockStatic
        }

        fun isWakeLockHeld(context: Context): Boolean {
            val lock = ModeService.getLock(context.applicationContext)
            return lock!!.isHeld
        }

        fun acquireWakeLock(context: Context) {
            val lock = ModeService.getLock(context.applicationContext)
            if (!lock!!.isHeld) {
                AppLog.d("WakeLock is not held")
                lock.acquire()
            }
            AppLog.d("WakeLock acquired")
        }

        fun releaseWakeLock(context: Context) {
            val lock = ModeService.getLock(context.applicationContext)

            if (lock!!.isHeld) {
                AppLog.d("WakeLock is held")
                lock.release()
            }
            AppLog.d("WakeLock released")
        }

        fun createStartIntent(context: Context, mode: Int): Intent {
            val service = Intent(context, ModeService::class.java)
            service.putExtra(ModeService.EXTRA_MODE, mode)
            return service
        }
    }
}
